### ISSUE TITLE
Critical: Normal people who are not using Automation scripts to view Slots, Are behind left behind. 

### ISSUE THREAD

**Issue by maheshwark97** [cowinapi/developer.cowin#193]
Please close down Public API's so that as to allow Vaccine Equality and a true first come first serve.
Even if a slot with 150 Available slots Is visible, Between the time difference when clicked on it and it is viewed, The Slot is already completely booked.

This is causing a huge equality issue as tech-savvy people(Especially Developers) are able to book slots faster.

**Comment by maheshwark97** [comment]
People are basically setting up temporary numbers in Android studio and using those numbers for registration and using Auto-Captcha fillers along with control via Public API and JavaScript automation to clear ahead of the general population. If the entire public API would be closed down, This would make things much easier for the general public who are not able to code every single step of the Process.

**Comment by saran-surya** [comment]
Hello @maheshwark97 , Yes I accept your comment, but most of the BOTS out there only provide data of the slots available, and the slots can only be booked through the cowin portal, And they have also restricted the access times to 100 calls per 5 minutes, Also the bots do not book appointments (As far as my knowledge), but notifying people about he slots availability, They just provide the data for the vaccine slots.

**Comment by gopal-amlekar** [comment]
> People are basically setting up temporary numbers in Android studio and using those numbers for registration and using Auto-Captcha fillers along with control via Public API and JavaScript automation to clear ahead of the general population. If the entire public API would be closed down, This would make things much easier for the general public who are not able to code every single step of the Process.

How do they work as some ID proof number is required for registration and also real mobile numbers with OTP? I also doubt that there is some bot-network which is booking slots immediately as they are available. But still can't get the idea how the modus operandi is? How the captcha can be bypassed? Booking appointment otherwise is only via protected APIs.

**Comment by maheshwark97** [comment]
Using Protected API's with token and exposing them using Man In the middle attacks on your own self to find out how they work and what kind of data is needed will solve the automation booking problem. If needed I will send you an entire brief of what is needed in the protected API's using Burp Suit as to how the booking is done. 
The second problem at hand is the OTP problem and using a number to get a booking. You can create a virtual Ennvironment in your windows system and emulate a temporary number to be redirected to it using an App on your phone so that OTP is instantly received. 

The third is checking Slots availability, You can use a server like AWS with multiple EC2 instances to solve this problem and have multiple IP addresses hitting at very fast speeds.

Closing down Public API's and securing Protected API's both are very much necessary.

**Comment by me-nkr** [comment]
@maheshwark97 The same public API is used by developers to provide notification to general public about slot openings.

**Comment by maheshwark97** [comment]
@me-nkr Is that fair that half the people are able to get notifications faster than the other half. And calling it half is being optimistic. I would say about 2-5% of the population is being pushed ahead of others in the race to get vaccinated because they are able to search on google better and know how to use notification services? That is unfair to the rest of the people.

**Comment by me-nkr** [comment]
@maheshwark97 I did think that before, but covid vaccines are given to a person one time ( one vaccine one time ). So someone booking it first means less people to get vaccinated and less delay between we getting vaccinated. Also the notification portals actually helps general people to get notified about slot opening. If someone is not using that utility and is complaining against the ones using it, whose fault is that ?

**Comment by gopal-amlekar** [comment]
> Using Protected API's with token and exposing them using Man In the middle attacks on your own self to find out how they work and what kind of data is needed will solve the automation booking problem. If needed I will send you an entire brief of what is needed in the protected API's using Burp Suit as to how the booking is done.
> The second problem at hand is the OTP problem and using a number to get a booking. You can create a virtual Ennvironment in your windows system and emulate a temporary number to be redirected to it using an App on your phone so that OTP is instantly received.
> 
> The third is checking Slots availability, You can use a server like AWS with multiple EC2 instances to solve this problem and have multiple IP addresses hitting at very fast speeds.
> 
> Closing down Public API's and securing Protected API's both are very much necessary.

Oh wow.. I thought captcha and OTP are just a nice security countermeasures. But this exploitation is not happening due to public APIs, it  is due to the security holes in the protected APIs then. So disabling public APIs will not help though. Also the protected APIs are directly visible in just the F12 tools anyway even if they had not published it on Setu.

I agree it is unfair use and it must be stopped somehow.

**Comment by saran-surya** [comment]
Yes @gopal-amlekar , They can implement some strict measures like generating proper API keys, for developers, And yes as you said all the API's are available directly on F12, thus they can easily scrape data....

**Comment by sprashant7** [comment]
I am 34 years old IT technologist staying in Mumbai and got my first jab on 9-May. I would like to share my experience. I was quite thrilled to see CoWIN portal as I could imagine what they have built, high performant AWS based horizontally scaling microservices based system, front end apis exposed through swagger etc. Hence I was very much inclined towards using CoWIN portal initially. Maybe due to that I used CoWIN portal from 28-Apr to 4-May but never got the confirmed appointment. During this period some of my friends got the confirmed appointment through Aarogya Setu. Hence I decided to switch to Aarogya Setu on 5-May, after trying for 4 days I got the appointment on 8-May (btw for the first three days I made stupid mistakes hence I should not count those days). On 9-May when I reached BKC vaccination center I did quick spot checks with nearly 8-9 people and asked them how did they book slots. All of them answered they used Aarogya Setu App.  Hence I reached to conclusion: **AAROGYA SETU GETS THE JAB DONE!**
**My Analysis of why Automated Scripts wont threaten equality because of two challenges:**
**Challenge1 (CAPTCHA)**: No matter which bot is how fast, it cannot crack captcha introduced by CoWIN in less than 10-20 seconds even with the good OCR libraries like Tesseract and 99% times it decodes it incorrectly. Hence, bots cannot automate this part.
**Challenge2 (OTP)**: While there are people who claim that they can forward OTP sms through IFTTT like apps, there are techies who understand its a risky thing to do as it exposes their SMS store to public cloud. 

**My Analysis of why AAROGYA SETU GETS THE JAB DONE!:**
**No Captcha**: I feel that this is the best thing which allows humans to win the race against any bots. Let bot creators burn their head, we will focus on few jabs for our-self and our family members.

**THE REAL PROBLEM**: The real problem is that there's no equilibrium in the equation. There's very high demand and very low supply of vaccines in the districts like Mumbai/Pune due to scarcity of vaccines. Until that problem is resolved lets focus on staying healthy and staying safe!

**Comment by dude83759** [comment]
Hey, how are Private APIs accessible through F12. Any screenshots?

> > Using Protected API's with token and exposing them using Man In the middle attacks on your own self to find out how they work and what kind of data is needed will solve the automation booking problem. If needed I will send you an entire brief of what is needed in the protected API's using Burp Suit as to how the booking is done.
> > The second problem at hand is the OTP problem and using a number to get a booking. You can create a virtual Ennvironment in your windows system and emulate a temporary number to be redirected to it using an App on your phone so that OTP is instantly received.
> > The third is checking Slots availability, You can use a server like AWS with multiple EC2 instances to solve this problem and have multiple IP addresses hitting at very fast speeds.
> > Closing down Public API's and securing Protected API's both are very much necessary.
> 
> Oh wow.. I thought captcha and OTP are just a nice security countermeasures. But this exploitation is not happening due to public APIs, it is due to the security holes in the protected APIs then. So disabling public APIs will not help though. Also the protected APIs are directly visible in just the F12 tools anyway even if they had not published it on Setu.
> 
> I agree it is unfair use and it must be stopped somehow.

**Comment by gopal-amlekar** [comment]
> Hey, how are Private APIs accessible through F12. Any screenshots?
> 
> > > Using Protected API's with token and exposing them using Man In the middle attacks on your own self to find out how they work and what kind of data is needed will solve the automation booking problem. If needed I will send you an entire brief of what is needed in the protected API's using Burp Suit as to how the booking is done.
> > > The second problem at hand is the OTP problem and using a number to get a booking. You can create a virtual Ennvironment in your windows system and emulate a temporary number to be redirected to it using an App on your phone so that OTP is instantly received.
> > > The third is checking Slots availability, You can use a server like AWS with multiple EC2 instances to solve this problem and have multiple IP addresses hitting at very fast speeds.
> > > Closing down Public API's and securing Protected API's both are very much necessary.
> > 
> > 
> > Oh wow.. I thought captcha and OTP are just a nice security countermeasures. But this exploitation is not happening due to public APIs, it is due to the security holes in the protected APIs then. So disabling public APIs will not help though. Also the protected APIs are directly visible in just the F12 tools anyway even if they had not published it on Setu.
> > I agree it is unfair use and it must be stopped somehow.

Not accessible, I said visible. You can see which APIs are being called but of course you can't get the API key and you can't use them on your own.

**Comment by bkrajendra** [comment]
I just wanted to add to all the above comments.
One thing really proved with this challenge! 
😎 We have a lot of talent who came out to solve this (booking) issue. (You can just search for Cowin on GitHub, there are a lot of open respos with slot availability and booking)

OTP auto verification does not need much effort, It can be very easily auto verified in microseconds using Cordova APIs,
or a simple user script in a browser with an SMS web app can do the job.

Technically there is no way to stop this as web apps are public-facing.
Captcha can just slow it down a little bit.
Cowin architecture is great, but the simple thing as adding captcha should have been done from day one, to stop people from writing codes to auto book.

Many smart developers might have stopped at the point when they see captcha as a security mechanism.
I know there are ways to break it but it's not worth writing code to just get the jab.
At least it could have made the process slow for coders.

**Comment by prashanthr84** [comment]
The whole system/process has to be changed. Instead of allowing people to book appointment, allow users to only register and let the system allot the slot to people randomly or any other criteria when vaccine is available. This is the same procedure followed in US(may be other countries).

**Comment by souvik234** [comment]
A) The ones who get notifications through Telegram groups, etc. don't have a real advantage. The only real way to win this game is to stay logged into Co-WIN, during the slot release times, and continuously keep refreshing. This anyone and everyone can do.

B) In cities with lakhs of people, those 150 slots are gonna get filled up in seconds irrespective of API or not. The API allows people a chance to have an idea of when to log-in to Co-WIN rather than this purely random game which you're suggesting.

**Comment by gopal-amlekar** [comment]
> A) The ones who get notifications through Telegram groups, etc. don't have a real advantage. The only real way to win this game is to stay logged into Co-WIN, during the slot release times, and continuously keep refreshing. This anyone and everyone can do.
> 
> B) In cities with lakhs of people, those 150 slots are gonna get filled up in seconds irrespective of API or not. The API allows people a chance to have an idea of when to log-in to Co-WIN rather than this purely random game which you're suggesting.

Well, how do you know the slot release times? I don't think there is any schedule to that. Every hospital probably takes their own call.

**Comment by souvik234** [comment]
@gopal-amlekar Yes. Every hospital has their own times, which can be easily figured out by looking at historical data from the Telegram groups/Twitter. That's what I meant by slot release times.

Though I do propose that there be a uniform set of slot release times followed by every vaccine center. Its better to have a set than just one, to prevent excessive load on the system.

**Comment by patelhir** [comment]
> Please close down Public API's so that as to allow Vaccine Equality and a true first come first serve.
> Even if a slot with 150 Available slots Is visible, Between the time difference when clicked on it and it is viewed, The Slot is already completely booked.
> 
> This is causing a huge equality issue as tech-savvy people(Especially Developers) are able to book slots faster.

I too feel same. I have written a simple python script to find available slots and book couple of slots (on Cowin portal) for myself, family and some other friends.  On co-win portal slots shows little late could be browser cache or time it takes to display HTML. Certainly it's difficult for non-techy to book a slot. With that in mind, I have decided to help now only high risk group people only.
